crates/arrow_sql_gen/src/postgres/composite.rs

```diff
@@ -0,0 +1,254 @@
+/*
+Copyright 2024 The Spice.ai OSS Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+     https://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+#![allow(clippy::module_name_repetitions)]
+
+use byteorder::{BigEndian, ByteOrder, ReadBytesExt};
+use fallible_iterator::FallibleIterator;
+use snafu::prelude::*;
+use std::{fmt, ops::Range};
+use tokio_postgres::{
+    row::RowIndex,
+    types::{Field, FromSql, Kind, Type, WrongType},
+};
+
+#[derive(Debug, Snafu)]
+pub enum Error {
+    #[snafu(display("Unable to parse composite type ranges: {source}"))]
+    UnableToParseCompositeTypeRanges { source: std::io::Error },
+
+    #[snafu(display("Unable to find column {column} in the fields {}", fields.join(", ")))]
+    UnableToFindColumnInFields { column: String, fields: Vec<String> },
+
+    #[snafu(display("{source}"))]
+    UnableToConvertType { source: WrongType },
+
+    #[snafu(display("Unable to conver raw bytes into expected type: {source}"))]
+    UnableToConvertBytesToType {
+        source: Box<dyn std::error::Error + Sync + Send>,
+    },
+}
+
+pub type Result<T, E = Error> = std::result::Result<T, E>;
+
+/// A `PostgreSQL` composite type.
+/// Fields of a type can be accessed using `CompositeType::get` and `CompositeType::try_get` methods.
+///
+/// Adapted from <https://github.com/sfackler/rust-postgres/pull/565>
```

Comment:
I took this entire file (with some small modifications to make lint happy) from this PR linked here.